clean-modular-code/checks-conditionals/python-function-checks.md

```diff
@@ -132,6 +165,58 @@ The EAFP (Easier to Ask for Forgiveness than Permission) approach is more Python
 
 As long as you consider edge cases, you're writing great code! You don’t need to worry about being “Pythonic” immediately, but understanding both approaches is useful regardless of which approach you chose.
 
++++ {"editable": true, "slideshow": {"slide_type": ""}}
+
+## Try/Except blocks
+
+Try/except blocks help you catch and handle errors that might happen while your code is running. This is useful for tasks that might fail, like converting data types or working with data that’s missing or incorrect.
+
+A try/except block looks like this:
+
+```python
+try:
+    # code that might cause an error
+except SomeError:
+    # what to do if there's an error
+```
+
+```{code-cell} ipython3
+---
+editable: true
+slideshow:
+  slide_type: ''
+---
+def convert_to_int(value):
+    try:
+        return int(value)
+    except ValueError:
+        print("Oops i can't process this so I will fail gracefully.")
```

Comment:
Nit: this isn't failing gracefully, this is recovering. Both are valid, depending on what guarantees the function wants to make. But I would probably not print any message if I was able to recover from an error and the rest of the program could continue

Reply:
@ucodery, I appreciate this. I was actually going to ask you about this. There are two sides here (I think?). If I raise the error, then I get the full stack output. My message is at the bottom. For a lot of people, that's a lot of "stuff" to read. 
<img width="908" alt="Screenshot 2024-10-11 at 5 09 30 PM" src="https://github.com/user-attachments/assets/d44cdba6-d4b5-4697-99a0-6a5d8a8dc164">

If i print the error - i understand that this is a "recovery"  as you call it above. from a user perspective (end user) it looks prettier / cleaner is easier to digest. 

<img width="911" alt="Screenshot 2024-10-11 at 5 10 41 PM" src="https://github.com/user-attachments/assets/c13ae17b-025f-4e5d-b03a-336240a052fd">

The program continues to run, which is problematic because, in this case, it will fail later because it has no data to process. From a development perspective, raising the error is nice for testing and feels more robust. it also should fail at this point for our demo workflow in this workshop because this means there is no data! But again the traceback output is a lot for a user to digest and has a lot of info that most users don't need. 

i'm curious what your thoughts are here in terms of what to suggest people do - or how do they make decisions?

After working through this, I was thinking about providing an easier example check around data that throws a key error or index error . That's easier because we'd want to keep processing but do something with that data to clean it even it that something is returning  `None`

Reply:
I can appreciate that printing the error message makes for a nicer intro to errors at runtime for students. I think my reaction to this is because `print` is in the except block, but then no error is raised. So the exception is passively squashed, and execution continues, without a `data` variable, possible causing issues later on (the opposite of "fail fast").

If you do (re)raise an exception, the student is now faced with the info dump of a stack trace as well as an error message. To avoid this, the program could `sys.exit()` or in some other way terminate.

---

As an aside, part of the reason there is so much going on in your first example is that the code is raising a new exception inside of an except block, which triggers a chained exception. This can be very useful to debugging, but to suppress the extra information that is not helpful in this case, `raise FileNotFoundError("Oops!") from None` can be used. 

This has the advantage of reducing unnecessary information, but makes teaching the `raise` and `except` statements potentially harder.

Another strategy, which adds complication to introducing basic error handling, but gives more control over the output, is to explicitly format exceptions. For instance, there is a stdlib that lets you do
```python
try:
    with open(file_path, "r") as file:
        data = file.read()
except FileNotFoundError as fe:
    print("".join(traceback.format_exception_only(fe)))
    sys.exit()
```
and there are 3rd party libraries that are meant to enhance tracebacks. Often this means adding more information, but there are ways to instead reduce the overall information ([like suppressing foreign libraries](https://rich.readthedocs.io/en/stable/traceback.html#suppressing-frames)).

Reply:
thank you @ucodery ok I think that I just learned a LOT more about exceptions. It is the chaining that I didn't fully understand. I know how to read the chains, but I didn't know about from None and about the concept of chaining. thank you. 

this lesson is actually a lot more complex that I originally thought it would be!